File: schemaorgphp/classes/reservation.php
<?php

class Reservation extends Intangible implements JsonSerializable {

	public static $context = 'http://schema.org';
	public static $type = 'Reservation';
	
	/**
	 * The date and time the reservation was booked.
	 * see : https://schema.org/bookingTime
	 * @var string|string[]
	 */
	public var $booking_time;
	
	/**
	 * An entity that arranges for an exchange between a buyer and a seller.  In most cases a broker never acquires or releases ownership of a product or service involved in an exchange.  If it is not clear whether an entity is a broker, seller, or buyer, the latter two terms are preferred. Supersedes bookingAgent (see: https://schema.org/bookingAgent).
	 * see : https://schema.org/broker
	 * @var \Organization|\Organization[]|\Person|\Person[]
	 */
	public var $broker;
	
	/**
	 * The date and time the reservation was modified.
	 * see : https://schema.org/modifiedTime
	 * @var string|string[]
	 */
	public var $modified_time;
	
	/**
	 * The currency (in 3-letter ISO 4217 format) of the price or a price component, when attached to PriceSpecification (see: https://schema.org/PriceSpecification) and its subtypes.
	 * see : https://schema.org/priceCurrency
	 * @var string|string[]
	 */
	public var $price_currency;
	
	/**
	 * Any membership in a frequent flyer, hotel loyalty program, etc. being applied to the reservation.
	 * see : https://schema.org/programMembershipUsed
	 * @var \ProgramMembership|\ProgramMembership[]
	 */
	public var $program_membership_used;
	
	/**
	 * The service provider, service operator, or service performer; the goods producer. Another party (a seller) may offer those services or goods on behalf of the provider. A provider may also serve as the seller. Supersedes carrier (see: https://schema.org/carrier).
	 * see : https://schema.org/provider
	 * @var \Organization|\Organization[]|\Person|\Person[]
	 */
	public var $provider;
	
	/**
	 * The thing -- flight, event, restaurant,etc. being reserved.
	 * see : https://schema.org/reservationFor
	 * @var \Thing|\Thing[]
	 */
	public var $reservation_for;
	
	/**
	 * A unique identifier for the reservation.
	 * see : https://schema.org/reservationId
	 * @var string|string[]
	 */
	public var $reservation_id;
	
	/**
	 * The current status of the reservation.
	 * see : https://schema.org/reservationStatus
	 * @var \ReservationStatusType|\ReservationStatusType[]
	 */
	public var $reservation_status;
	
	/**
	 * A ticket associated with the reservation.
	 * see : https://schema.org/reservedTicket
	 * @var \Ticket|\Ticket[]
	 */
	public var $reserved_ticket;
	
	/**
	 * The total price for the reservation or ticket, including applicable taxes, shipping, etc.
	 * see : https://schema.org/totalPrice
	 * @var float|float[]|\PriceSpecification|\PriceSpecification[]|string|string[]
	 */
	public var $total_price;
	
	/**
	 * The person or organization the reservation or ticket is for.
	 * see : https://schema.org/underName
	 * @var \Organization|\Organization[]|\Person|\Person[]
	 */
	public var $under_name;
	
	public function jsonSerialize() {
		$out = array(
			'@context' => 'http://schema.org',
			'@type' => 'Reservation'
		);
		
		$serialized = so_json_serialize( $this->booking_time );
		if ( ! empty( $serialized ) ) {
			$out['bookingTime'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->broker );
		if ( ! empty( $serialized ) ) {
			$out['broker'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->modified_time );
		if ( ! empty( $serialized ) ) {
			$out['modifiedTime'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->price_currency );
		if ( ! empty( $serialized ) ) {
			$out['priceCurrency'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->program_membership_used );
		if ( ! empty( $serialized ) ) {
			$out['programMembershipUsed'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->provider );
		if ( ! empty( $serialized ) ) {
			$out['provider'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->reservation_for );
		if ( ! empty( $serialized ) ) {
			$out['reservationFor'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->reservation_id );
		if ( ! empty( $serialized ) ) {
			$out['reservationId'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->reservation_status );
		if ( ! empty( $serialized ) ) {
			$out['reservationStatus'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->reserved_ticket );
		if ( ! empty( $serialized ) ) {
			$out['reservedTicket'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->total_price );
		if ( ! empty( $serialized ) ) {
			$out['totalPrice'] = $serialized;
		}
		
		$serialized = so_json_serialize( $this->under_name );
		if ( ! empty( $serialized ) ) {
			$out['underName'] = $serialized;
		}
		
		return $out;
	}
}

File: schemaorggo/reservation.go
package schemaorggo

import "encoding/json"

// Reservation see : https://schema.org/Reservation
type Reservation struct {
	Intangible

	typeContext

	// BookingTime see : https://schema.org/bookingTime
	// The date and time the reservation was booked.
	// types : DateTime
	BookingTime []DateTime `json:"bookingTime,omitempty"`

	// Broker see : https://schema.org/broker
	// An entity that arranges for an exchange between a buyer and a seller.  In most cases a broker never acquires or releases ownership of a product or service involved in an exchange.  If it is not clear whether an entity is a broker, seller, or buyer, the latter two terms are preferred. Supersedes bookingAgent (see: https://schema.org/bookingAgent).
	// types : Organization Person
	Broker []interface{} `json:"broker,omitempty"`

	// ModifiedTime see : https://schema.org/modifiedTime
	// The date and time the reservation was modified.
	// types : DateTime
	ModifiedTime []DateTime `json:"modifiedTime,omitempty"`

	// PriceCurrency see : https://schema.org/priceCurrency
	// The currency (in 3-letter ISO 4217 format) of the price or a price component, when attached to PriceSpecification (see: https://schema.org/PriceSpecification) and its subtypes.
	// types : Text
	PriceCurrency []string `json:"priceCurrency,omitempty"`

	// ProgramMembershipUsed see : https://schema.org/programMembershipUsed
	// Any membership in a frequent flyer, hotel loyalty program, etc. being applied to the reservation.
	// types : ProgramMembership
	ProgramMembershipUsed []*ProgramMembership `json:"programMembershipUsed,omitempty"`

	// Provider see : https://schema.org/provider
	// The service provider, service operator, or service performer; the goods producer. Another party (a seller) may offer those services or goods on behalf of the provider. A provider may also serve as the seller. Supersedes carrier (see: https://schema.org/carrier).
	// types : Organization Person
	Provider []interface{} `json:"provider,omitempty"`

	// ReservationFor see : https://schema.org/reservationFor
	// The thing -- flight, event, restaurant,etc. being reserved.
	// types : Thing
	ReservationFor []*Thing `json:"reservationFor,omitempty"`

	// ReservationId see : https://schema.org/reservationId
	// A unique identifier for the reservation.
	// types : Text
	ReservationId []string `json:"reservationId,omitempty"`

	// ReservationStatus see : https://schema.org/reservationStatus
	// The current status of the reservation.
	// types : ReservationStatusType
	ReservationStatus []*ReservationStatusType `json:"reservationStatus,omitempty"`

	// ReservedTicket see : https://schema.org/reservedTicket
	// A ticket associated with the reservation.
	// types : Ticket
	ReservedTicket []*Ticket `json:"reservedTicket,omitempty"`

	// TotalPrice see : https://schema.org/totalPrice
	// The total price for the reservation or ticket, including applicable taxes, shipping, etc.
	// types : Number PriceSpecification Text
	TotalPrice []interface{} `json:"totalPrice,omitempty"`

	// UnderName see : https://schema.org/underName
	// The person or organization the reservation or ticket is for.
	// types : Organization Person
	UnderName []interface{} `json:"underName,omitempty"`
}

func (v Reservation) intoMap(intop *map[string]interface{}) error {
	if intop == nil {
		return nil
	}

	v.Intangible.intoMap(intop)

	into := *intop

	{
		var value interface{} = v.BookingTime
		if len(v.BookingTime) == 1 {
			value = v.BookingTime[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["bookingTime"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.Broker
		if len(v.Broker) == 1 {
			value = v.Broker[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["broker"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.ModifiedTime
		if len(v.ModifiedTime) == 1 {
			value = v.ModifiedTime[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["modifiedTime"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.PriceCurrency
		if len(v.PriceCurrency) == 1 {
			value = v.PriceCurrency[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["priceCurrency"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.ProgramMembershipUsed
		if len(v.ProgramMembershipUsed) == 1 {
			value = v.ProgramMembershipUsed[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["programMembershipUsed"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.Provider
		if len(v.Provider) == 1 {
			value = v.Provider[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["provider"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.ReservationFor
		if len(v.ReservationFor) == 1 {
			value = v.ReservationFor[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["reservationFor"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.ReservationId
		if len(v.ReservationId) == 1 {
			value = v.ReservationId[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["reservationId"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.ReservationStatus
		if len(v.ReservationStatus) == 1 {
			value = v.ReservationStatus[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["reservationStatus"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.ReservedTicket
		if len(v.ReservedTicket) == 1 {
			value = v.ReservedTicket[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["reservedTicket"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.TotalPrice
		if len(v.TotalPrice) == 1 {
			value = v.TotalPrice[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["totalPrice"] = json.RawMessage(b)
		}
	}

	{
		var value interface{} = v.UnderName
		if len(v.UnderName) == 1 {
			value = v.UnderName[0]
		}

		b, err := json.Marshal(value)
		if err != nil {
			return err
		}

		if len(b) > 0 && string(b) != "null" {
			into["underName"] = json.RawMessage(b)
		}
	}

	*intop = into

	return nil
}

func (v Reservation) asMap() (map[string]interface{}, error) {
	data := map[string]interface{}{}
	err := v.intoMap(&data)
	if err != nil {
		return nil, err
	}

	data["@context"] = "http://schema.org"
	data["@type"] = "Reservation"

	return data, nil
}

func (v Reservation) MarshalJSON() ([]byte, error) {
	data, err := v.asMap()
	if err != nil {
		return nil, err
	}

	return json.Marshal(data)
}
